pam/internal/adapter/gdmmodel_test.go

```diff
@@ -0,0 +1,2343 @@
+					},
+				},
+			},
+		}, nil
+
+	case *gdm.RequestData_ChangeStage:
+		h.t.Logf("Switching to stage %s", req.ChangeStage.Stage)
+		h.stageChanges = append(h.stageChanges, req.ChangeStage.Stage)
+
+		h.currentStage = req.ChangeStage.Stage
+		h.currentStageCond.L.Unlock()
+		h.currentStageCond.Broadcast()
+		h.currentStageCond.L.Lock()
+
+		return &gdm.Data{
+			Type: gdm.DataType_response,
+			Response: &gdm.ResponseData{
+				Type: gdmData.Request.Type,
+				Data: &gdm.ResponseData_Ack{},
+			},
+		}, nil
+
+	default:
+		return nil, fmt.Errorf("unknown request type")
+	}
+}
+
+func (h *gdmConvHandler) handleEvent(event *gdm.EventData) error {
+	defer func() {
+		h.receivedEvents = append(h.receivedEvents, event)
+
+		if h.wantEvents == nil {
+			return
+		}
+		if !h.checkAllEventsHaveBeenEmitted() {
+			return
+		}
+
+		h.wantEvents = nil
+		go func() {
+			// Mark the events received after or while we're returning.
+			close(h.allEventsReceived)
+		}()
+	}()
+
+	switch ev := event.Data.(type) {
+	case *gdm.EventData_BrokersReceived:
+		h.receivedBrokers = ev.BrokersReceived.BrokersInfos
+
+	case *gdm.EventData_BrokerSelected:
+		h.selectedBrokerID = ev.BrokerSelected.BrokerId
+
+	case *gdm.EventData_AuthModesReceived:
+		// TODO: Check the auth modes are matching.
+
+	case *gdm.EventData_UiLayoutReceived:
+		if !slices.ContainsFunc(h.supportedLayouts, func(layout *authd.UILayout) bool {
+			return layout.Type == ev.UiLayoutReceived.UiLayout.Type
+		}) {
+			return fmt.Errorf(`unknown layout type: "%s"`, ev.UiLayoutReceived.UiLayout.Type)
+		}
+	}
+
+	return nil
+}
+
+func (h *gdmConvHandler) waitForStageChange(stage proto.Stage) func() {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	if h.currentStage == stage && (h.lastNotifiedStage == nil || *h.lastNotifiedStage != stage) {
+		h.lastNotifiedStage = &stage
+		return nil
+	}
+
+	return func() {
+		h.currentStageCond.L.Lock()
+		defer h.currentStageCond.L.Unlock()
+
+		for {
+			// We just got notified of a stage change but we should not notify all the waiting
+			// waiting requests all together, each request to this function should be queued.
+			// So the goroutine that won the lock is the one that will be blocked if the stage
+			// matches and if it's the first one noticing such change.
+			if h.currentStage == stage && (h.lastNotifiedStage == nil || *h.lastNotifiedStage != stage) {
+				h.lastNotifiedStage = &stage
+				return
+			}
+
+			h.currentStageCond.Wait()
+		}
+	}
+}
+
+func (h *gdmConvHandler) appendPollResultEvents(events ...*gdm.EventData) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	h.pendingEvents = append(h.pendingEvents, events...)
+}
+
+// Custom messages for testing the gdm model.
+
+type testGdmAddPollResultEvent struct {
+	event *gdm.EventData
+}
+
+type testGdmWaitForStage struct {
+	stage    proto.Stage
+	events   []*gdm.EventData
+	commands []tea.Cmd
+}
+
+type testGdmWaitForStageDone testGdmWaitForStage
+
+type testGdmWaitForStageCommandsDone struct {
```

Comment:
I would also move these and their methods to a separate file to avoid cluttering the actual tests.